EventFlow.Redis/EventStore/RedisEventPersistence.cs

```diff
@@ -0,0 +1,162 @@
+    }
+
+    public async Task<AllCommittedEventsPage> LoadAllCommittedEvents(GlobalPosition globalPosition, int pageSize,
+        CancellationToken cancellationToken)
+    {
+        var startPosition = globalPosition.IsStart
+            ? 0
+            : long.Parse(globalPosition.Value);
+
+        var streamNames = await _resolver.GetStreamIdsAsync(cancellationToken).ConfigureAwait(false);
+        var streamTasks = streamNames.Select(prefixedKey =>
+            GetCommittedEventsAsync(prefixedKey, startPosition, cancellationToken, pageSize)).ToList();
+
+        await Task.WhenAll(streamTasks).ConfigureAwait(false);
+        var events = streamTasks.SelectMany(t => t.Result);
+
+        var nextPos = events.Any()
+            ? events.Max(e => e.AggregateSequenceNumber)
+            : startPosition;
+
+        if (_logger.IsEnabled(LogLevel.Trace))
+        {
+            _logger.LogTrace("Loaded {Count} events from redis", events.Count());
+        }
+
+        return new AllCommittedEventsPage(new GlobalPosition(nextPos.ToString()), events.ToList());
+    }
+
+    public async Task<IReadOnlyCollection<ICommittedDomainEvent>> CommitEventsAsync(IIdentity id,
+        IReadOnlyCollection<SerializedEvent> serializedEvents, CancellationToken cancellationToken)
+    {
+        var committedEvents = new List<RedisCommittedDomainEvent>();
+        var database = _multiplexer.GetDatabase();
+        var prefixedKey = GetAsPrefixedKey(id.Value);
+
+        foreach (var serializedEvent in serializedEvents)
+        {
+            //Redis stream entry id uses the format: <UnixTime>-<IncrementingId>, we leave the timestamp blank to ensure optimistic concurrency works
+            var messageId = $"0-{serializedEvent.AggregateSequenceNumber}";
+
+            var data = new NameValueEntry("data", serializedEvent.SerializedData);
+            var metadata = new NameValueEntry("metadata", serializedEvent.SerializedMetadata);
+
+            try
+            {
+                var result = await database.StreamAddAsync(prefixedKey, new[] {data, metadata}, messageId)
+                    .ConfigureAwait(false);
+                if (result == messageId)
+                {
+                    if (_logger.IsEnabled(LogLevel.Trace))
+                        _logger.LogTrace(
+                            "Committed event with id {EventId} for aggregate with Id {AggregateId} to Redis ",
+                            messageId, prefixedKey.Key);
+
+                    committedEvents.Add(new RedisCommittedDomainEvent(prefixedKey.Key, data.Value, metadata.Value,
+                        serializedEvent.AggregateSequenceNumber));
+                }
+                else
+                {
+                    _logger.LogWarning(
+                        "Failed to commit event with id {EventId} for aggregate with Id {AggregateId} to Redis",
+                        prefixedKey.Key, messageId);
+
+                    throw new Exception(result);
```

Comment:
Can you make this exception more useful? or can it be transient? Maybe a code comment on when this can actually happen?

Reply:
After taking a look at the redis and redis stackexchange documentation, this should never happen. Removed the exception